pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,150 +18,328 @@ import (
-		topologyGroups = append(topologyGroups, topologyGroup)
+	// for anti-affinities, we record where the pods could be, even if
+	// requirements haven't collapsed to a single value.
+	for _, tc := range t.inverseTopologies {
+		if tc.IsOwnedBy(p.UID) {
+			tc.Record(requirements.Get(tc.Key).Values().UnsortedList()...)
+		}
 	}
-	return topologyGroups
 }
 
-func (t *Topology) computeCurrentTopology(ctx context.Context, constraints *v1alpha5.Constraints, topologyGroup *TopologyGroup) error {
-	switch topologyGroup.Constraint.TopologyKey {
-	case v1.LabelHostname:
-		return t.computeHostnameTopology(topologyGroup, constraints)
-	case v1.LabelTopologyZone:
-		return t.computeZonalTopology(ctx, constraints, topologyGroup)
-	case v1alpha5.LabelCapacityType:
-		return t.computeLabelCapacity(ctx, constraints, topologyGroup)
-	default:
-		return nil
-	}
-}
+// Requirements tightens the input requirements by adding additional requirements that are being enforced by topology spreads
+// affinities, anti-affinities or inverse anti-affinities.  It returns these newly tightened requirements, or an error in
+// the case of a set of requirements that cannot be satisfied.
+func (t *Topology) Requirements(requirements v1alpha5.Requirements, nodeHostname string, p *v1.Pod) (v1alpha5.Requirements, error) {
+	for _, topology := range t.getMatchingTopologies(p) {
+		// the node name may be brand new and unknown to the topology, so we need to ensure that the topology
+		// is aware that it's a valid hostname
+		if topology.Key == v1.LabelHostname {
+			topology.Register(nodeHostname)
+		}
 
-// computeHostnameTopology for the topology group. Hostnames are guaranteed to
-// be unique when new nodes join the cluster. Nodes that join the cluster do not
-// contain any pods, so we can assume that the global minimum domain count for
-// `hostname` is 0. Thus, we can always improve topology skew (computed against
-// the global minimum) by adding pods to the cluster. We will generate
-// len(pods)/MaxSkew number of domains, to ensure that skew is not violated for
-// new instances.
-func (t *Topology) computeHostnameTopology(topologyGroup *TopologyGroup, constraints *v1alpha5.Constraints) error {
-	domains := []string{}
-	for i := 0; i < int(math.Ceil(float64(len(topologyGroup.Pods))/float64(topologyGroup.Constraint.MaxSkew))); i++ {
-		domains = append(domains, strings.ToLower(randomdata.Alphanumeric(8)))
-	}
-	topologyGroup.Register(domains...)
-	// This is a bit of a hack that allows the constraints to recognize viable hostname topologies
-	constraints.Requirements = constraints.Requirements.Add(v1.NodeSelectorRequirement{Key: topologyGroup.Constraint.TopologyKey, Operator: v1.NodeSelectorOpIn, Values: domains})
-	return nil
+		nextDomain, err := topology.Next(requirements, topology.Matches(p.Namespace, p.Labels))
+		if err != nil {
+			return v1alpha5.Requirements{}, err
+		}
+
+		requirements = requirements.Add(v1.NodeSelectorRequirement{
```

Comment:
Can we return a set of requirements without mutating the passed in parameter and do the mutation outside?

Reply:
Leaving comments here that we talked about for reminder/reference:

It doesn't mutate the incoming requirements, .Add() returns a new set of requirements that we keep adding to as we further tighten the requirements by topology.  We need to keep using this ever-tightening set of requirements as we have to provide it to the topology Next() method so it's aware of the valid domains that it can choose from.